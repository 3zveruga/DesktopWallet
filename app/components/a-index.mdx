---
name: Home
route: /
---

# UI Components

Use the sidebar menu to browse the styleguide of React components available for the Zepio application.
