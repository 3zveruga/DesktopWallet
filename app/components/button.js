// @flow

import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable max-len */
// $FlowFixMe
import { darken } from 'polished';

const DefaultButton = styled.button`
  padding: 10px 30px;
  font-family: ${// $FlowFixMe
  props => props.theme.fontFamily};
  font-weight: ${// $FlowFixMe
  props => props.theme.fontWeight.bold};
  font-size: ${// $FlowFixMe
  props => `${props.theme.fontSize.text}em`};
  cursor: pointer;
  outline: none;
  min-width: 100px;
  border-radius: 100px;
  transition: background-color 0.1s ease-in-out;
  width: 100%;
`;

const Primary = styled(DefaultButton)`
  background-color: ${props => props.theme.colors.primary};
  color: ${props => props.theme.colors.secondary};
  border: none;

  &:hover {
    background-color: ${props => darken(0.1, props.theme.colors.primary(props))};
  }

  &:disabled {
    background-color: #3e3c42;
    cursor: not-allowed;
    opacity: 0.8;
  }
`;

const Secondary = styled(DefaultButton)`
  background-color: transparent;
  color: ${props => props.theme.colors.secondary};
  border: 2px solid #3e3c42;

  &:hover {
    border-color: ${props => props.theme.colors.primary};
  }

  &:disabled {
    background-color: Transparent;
    cursor: not-allowed;
    color: #3e3c42;

    &:hover {
      border-color: #3e3c42;
    }
  }
`;

type Props = {
  label: string,
  onClick?: () => void,
  to?: string,
  variant?: 'primary' | 'secondary',
  disabled?: boolean,
  className?: string,
};

export const Button = ({
  onClick,
  label,
  to,
  variant,
  disabled,
  className,
}: Props) => {
  if (to && onClick) throw new Error('Should define either "to" or "onClick"');

  const component = variant === 'primary' ? (
    <Primary onClick={onClick} disabled={disabled} className={className}>
      {label}
    </Primary>
  ) : (
    <Secondary onClick={onClick} disabled={disabled} className={className}>
      {label}
    </Secondary>
  );

  return to ? <Link to={String(to)}>{component}</Link> : component;
};

Button.defaultProps = {
  to: null,
  variant: 'primary',
  onClick: null,
  disabled: false,
  className: '',
};
