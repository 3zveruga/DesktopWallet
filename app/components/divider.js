// @flow
import styled from 'styled-components';

export const Divider = styled.div`
  width: 100%;
  height: 1px;
  background-color: ${props => props.color || props.theme.colors.text};
  opacity: ${props => props.opacity || 1};
  margin-bottom: ${props => props.marginBottom || 0};
  margin-top: ${props => props.marginTop || 0};
`;
