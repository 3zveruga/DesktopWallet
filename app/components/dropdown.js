// @flow
import React, { type Node, Component } from 'react';
import styled from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
// $FlowFixMe
import { darken } from 'polished';
import Popover from 'react-popover';
import ClickOutside from 'react-click-outside';

import { TextComponent } from './text';

/* eslint-disable max-len  */
const MenuWrapper = styled.div`
  background-image: ${props => `linear-gradient(to right, ${darken(
    0.05,
    props.theme.colors.activeItem,
  )}, ${props.theme.colors.activeItem})`};
  padding: 10px 0;
  border-radius: 10px;
  margin-left: -10px;
`;

const MenuItem = styled.button`
  outline: none;
  background-color: transparent;
  border: none;
  border-bottom-style: solid;
  border-bottom-color: ${props => props.theme.colors.text};
  border-bottom-width: 1px;
  padding: 15px 10px;
  cursor: pointer;
  width: 100%;
  text-align: left;

  &:hover {
    opacity: 0.9;
  }

  &:disabled {
    cursor: default;

    &:hover {
      opacity: 1;
    }
  }
`;

const PopoverWithStyle = styled(Popover)`
  & > .Popover-tip {
    fill: ${props => props.theme.colors.activeItem};
  }
`;

type Props = {
  renderTrigger: (toggleVisibility: () => void, isOpen: boolean) => Node,
  options: Array<{ label: string, onClick: () => void }>,
  label?: string | null,
};

type State = {
  isOpen: boolean,
};

export class DropdownComponent extends Component<Props, State> {
  state = {
    isOpen: false,
  };

  static defaultProps = {
    label: null,
  };

  render() {
    const { isOpen } = this.state;
    const { label, options, renderTrigger } = this.props;

    return (
      <PopoverWithStyle
        isOpen={isOpen}
        preferPlace='below'
        enterExitTransitionDurationMs={0}
        body={[
          <ClickOutside
            onClickOutside={() => this.setState(() => ({ isOpen: false }))}
          >
            <MenuWrapper>
              {label && (
                <MenuItem disabled>
                  <TextComponent value={label} isBold />
                </MenuItem>
              )}
              {options.map(({ label: optionLabel, onClick }) => (
                <MenuItem onClick={onClick}>
                  <TextComponent value={optionLabel} />
                </MenuItem>
              ))}
            </MenuWrapper>
          </ClickOutside>,
        ]}
        tipSize={7}
      >
        {renderTrigger(
          () => this.setState(state => ({ isOpen: !state.isOpen })),
          isOpen,
        )}
      </PopoverWithStyle>
    );
  }
}
