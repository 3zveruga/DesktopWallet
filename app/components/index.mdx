---
name: Home
route: /
---

# UI Components

Use the sidebar menu to browse the styleguide
