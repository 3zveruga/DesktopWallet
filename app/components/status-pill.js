// @flow
import React, { Component } from 'react';
import styled, { keyframes } from 'styled-components';
import eres from 'eres';

import { TextComponent } from './text';

import rpc from '../../services/api';

import readyIcon from '../assets/images/green_check.png';
import syncIcon from '../assets/images/sync_icon.png';
import errorIcon from '../assets/images/error_icon.png';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

const Wrapper = styled.div`
  align-items: center;
  display: flex;
  background-color: #000;
  border-radius: 27px;
  padding: 8px 16px;
`;

const Icon = styled.img`
  width: 12px;
  height: 12px;
  margin-right: 8px;
  animation: 2s linear infinite;
  animation-name: ${props => (props.animated ? rotate : 'none')};
`;

const StatusPillLabel = styled(TextComponent)`
  color: ${props => props.theme.colors.statusPillLabel};
  font-weight: ${props => props.theme.fontWeight.bold};
  text-transform: uppercase;
  font-size: 10px;
  padding-top: 1px;
  user-select: none;
`;

type Props = {};

type State = {
  type: string,
  icon: string,
  progress: number,
  isSyncing: boolean,
};

export class StatusPill extends Component<Props, State> {
  timer: ?IntervalID = null;

  state = {
    type: 'syncing',
    icon: syncIcon,
    progress: 0,
    isSyncing: true,
  };

  componentDidMount() {
    this.timer = setInterval(() => {
      this.getBlockchainStatus();
    }, 2000);
  }

  componentWillUnmount() {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  getBlockchainStatus = async () => {
    const [blockchainErr, blockchaininfo] = await eres(rpc.getblockchaininfo());

    if (blockchainErr || !blockchaininfo) return;

    const newProgress = blockchaininfo.verificationprogress * 100;

    this.setState({
      progress: newProgress,
      ...(newProgress > 99.99
        ? {
          type: 'ready',
          icon: readyIcon,
          isSyncing: false,
        }
        : {}),
    });

    if (blockchainErr) {
      this.setState(() => ({ type: 'error', icon: errorIcon }));
    }
  };

  render() {
    const {
      type, icon, progress, isSyncing,
    } = this.state;
    const showPercent = isSyncing ? `(${progress.toFixed(2)}%)` : '';

    return (
      <Wrapper id='status-pill'>
        <Icon src={icon} animated={isSyncing} />
        <StatusPillLabel value={`${type} ${showPercent}`} />
      </Wrapper>
    );
  }
}
