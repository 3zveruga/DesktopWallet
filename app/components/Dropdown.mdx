---
name: DropDown
---

import { Playground, PropsTable } from 'docz'

import { Dropdown } from './dropdown.js'
import { DoczWrapper } from '../theme.js'

# DropDown

<PropsTable of={Dropdown} />

## Basic usage

<Playground>
  <DoczWrapper>
    {() => (
      <div style={{ height: '500px' }}>
        <Dropdown
          options={[
            { label: 'asbh1yeasbdh23848asdasd', onClick: console.log },
            { label: 'urtyruhjr374hbfdjdhuh', onClick: console.log },
          ]}
          renderTrigger={toggleVisibility => <button onClick={toggleVisibility}>Opa</button>}
        />
      </div>
    )}
  </DoczWrapper>
</Playground>
