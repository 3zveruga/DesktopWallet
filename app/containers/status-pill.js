// @flow

import { connect } from 'react-redux';
import eres from 'eres';
import { BigNumber } from 'bignumber.js';
import { updateNodeSyncStatus } from '../redux/modules/app';

import { StatusPill } from '../components/status-pill';

import rpc from '../../services/api';

import type { Dispatch } from '../types/redux';
import type { AppState } from '../types/app-state';

export type MapStateToProps = {|
  nodeSyncProgress: number,
  nodeSyncType: 'ready' | 'syncing' | 'error',
|};

const mapStateToProps = ({ app }: AppState): MapStateToProps => ({
  nodeSyncProgress: app.nodeSyncProgress,
  nodeSyncType: app.nodeSyncType,
});

export type MapDispatchToProps = {|
  getBlockchainStatus: () => Promise<void>,
|};

const mapDispatchToProps = (dispatch: Dispatch): MapDispatchToProps => ({
  getBlockchainStatus: async () => {
    const [blockchainErr, blockchaininfo] = await eres(rpc.getblockchaininfo());

    if (blockchainErr || !blockchaininfo) {
      return dispatch(
        updateNodeSyncStatus({
          nodeSyncProgress: 0,
          nodeSyncType: 'error',
        }),
      );
    }

    const newProgress = blockchaininfo.verificationprogress * 100;

    dispatch(
      updateNodeSyncStatus({
        nodeSyncProgress: newProgress,
        nodeSyncType: new BigNumber(newProgress).gt(99.99) ? 'ready' : 'syncing',
      }),
    );
  },
});

// $FlowFixMe
export const StatusPillContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(StatusPill);
