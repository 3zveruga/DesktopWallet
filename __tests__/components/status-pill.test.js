// @flow

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import 'jest-dom/extend-expect';

import { StatusPill } from '../../app/components/status-pill';
import appTheme from '../../app/theme';

afterEach(cleanup);

describe('<StatusPill />', () => {
  describe('render()', () => {
    test('should render visual component status correctly', () => {
      // $FlowFixMe
      const { asFragment } = render(
        <ThemeProvider theme={appTheme}>
          <MemoryRouter>
            <StatusPill progress={83.} type='syncing' />
          </MemoryRouter>
        </ThemeProvider>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
