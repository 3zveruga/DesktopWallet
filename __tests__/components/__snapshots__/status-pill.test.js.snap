// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatusPill /> render() should render visual component status correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bwzfXH klEXgT"
  >
    <img
      class="sc-htpNat geYqkR"
      src="[object Object]"
    />
    <p
      class="sc-bxivhb gsACzh sc-bdVaJa fMeavL"
      color="#FFF"
    >
      syncing (83.00%)
    </p>
  </div>
</DocumentFragment>
`;
